lib/services/local/vnet_config.go

```diff
@@ -0,0 +1,121 @@
+	vnetConfigSingletonName = "vnet-config"
+)
+
+// VnetConfigService implements the storage layer for the VnetConfig resource.
+type VnetConfigService struct {
+	slog *slog.Logger
+	svc  *generic.ServiceWrapper[*vnet.VnetConfig]
+}
+
+// NewVnetConfigService returns a new VnetConfig storage service.
+func NewVnetConfigService(backend backend.Backend) (*VnetConfigService, error) {
+	svc, err := generic.NewServiceWrapper(
+		backend,
+		types.KindVnetConfig,
+		vnetConfigPrefix,
+		services.MarshalProtoResource[*vnet.VnetConfig],
+		services.UnmarshalProtoResource[*vnet.VnetConfig],
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &VnetConfigService{
+		svc:  svc,
+		slog: slog.With(teleport.ComponentKey, "VnetConfig.local"),
+	}, nil
+}
+
+// GetVnetConfig returns the singleton VnetConfig resource.
+func (s *VnetConfigService) GetVnetConfig(ctx context.Context) (*vnet.VnetConfig, error) {
+	vnetConfig, err := s.svc.GetResource(ctx, vnetConfigSingletonName)
+	return vnetConfig, trace.Wrap(err)
+}
+
+// CreateVnetConfig does basic validation and creates a VnetConfig resource.
+func (s *VnetConfigService) CreateVnetConfig(ctx context.Context, vnetConfig *vnet.VnetConfig) (*vnet.VnetConfig, error) {
+	if err := validateVnetConfig(vnetConfig); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	vnetConfig, err := s.svc.CreateResource(ctx, vnetConfig)
+	return vnetConfig, trace.Wrap(err)
+}
+
+// UpdateVnetConfig does basic validation and updates a VnetConfig resource.
+func (s *VnetConfigService) UpdateVnetConfig(ctx context.Context, vnetConfig *vnet.VnetConfig) (*vnet.VnetConfig, error) {
+	if err := validateVnetConfig(vnetConfig); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	vnetConfig, err := s.svc.ConditionalUpdateResource(ctx, vnetConfig)
+	return vnetConfig, trace.Wrap(err)
+}
+
+// UpsertVnetConfig does basic validation and upserts a VnetConfig resource.
+func (s *VnetConfigService) UpsertVnetConfig(ctx context.Context, vnetConfig *vnet.VnetConfig) (*vnet.VnetConfig, error) {
+	if err := validateVnetConfig(vnetConfig); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	vnetConfig, err := s.svc.UpsertResource(ctx, vnetConfig)
+	return vnetConfig, trace.Wrap(err)
+}
+
+// DeleteVnetConfig deletes the singleton VnetConfig resource.
+func (s *VnetConfigService) DeleteVnetConfig(ctx context.Context) error {
+	return trace.Wrap(s.svc.DeleteResource(ctx, vnetConfigSingletonName))
+}
+
+func validateVnetConfig(vnetConfig *vnet.VnetConfig) error {
+	if vnetConfig.GetKind() != types.KindVnetConfig {
+		return trace.BadParameter("kind must be %q", types.KindVnetConfig)
+	}
+	if vnetConfig.GetVersion() != types.V1 {
+		return trace.BadParameter("version must be %q", types.V1)
+	}
+	if vnetConfig.GetMetadata().GetName() != vnetConfigSingletonName {
+		return trace.BadParameter("name must be %q", vnetConfigSingletonName)
+	}
+	if cidrRange := vnetConfig.GetSpec().GetIpv4CidrRange(); cidrRange != "" {
```

Comment:
we should validate the dns zone suffixes as well to make sure they are valid

Reply:
sure, added some basic validation